File: color_matching.py
# color_matching.py

import numpy as np
import cv2 as cv

# Function that tries to find red areas in an image img
# Returns image, found, top_left, bottom_right where:
# image = the image, with a rectangle drawn around the found area
# found = true if a red area has been found
# top_left, bottom_right = the top left and bottom right corners
# of the found area
def color_matching(img):
    # Initialize default values
    MIN_RADIUS = 8
    bbox = [0, 0, 0, 0]
    found = False

    # Blur image
    img = cv.GaussianBlur(img, (5, 5), 0)

    # Change image to HSV colour space
    hsv = cv.cvtColor(img, cv.COLOR_BGR2HSV)

    # Mask out the lower interval
    lower_red = np.array([0, 130, 120])
    upper_red = np.array([10, 255, 255])
    mask1 = cv.inRange(hsv, lower_red, upper_red)

    # Mask out the upper interval
    lower_red = np.array([170, 120, 120])
    upper_red = np.array([180, 255, 255])
    mask2 = cv.inRange(hsv, lower_red, upper_red)

    # Add the two masks
    mask = mask1 + mask2

    # Find contours
    cnts, _ = cv.findContours(mask.copy(), cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)

    if len(cnts) > 0:
        # Pick the largest contour
        c = max(cnts, key=cv.contourArea)

        # Find the smallest enclosing circle
        _, radius = cv.minEnclosingCircle(c)
        # Find the bounding rectangle
        bbox = cv.boundingRect(c)

        if radius > MIN_RADIUS:
            found = True
            cv.rectangle(img, (bbox[0], bbox[1]), (bbox[0] + bbox[2], bbox[1] + bbox[3]), (0, 255, 0), 3)
    return img, found, (bbox[0], bbox[1]), (bbox[0] + bbox[2], bbox[1] + bbox[3])        
  

   
    

File: color_matching.cc
#include <opencv2/opencv.hpp>
#include <iostream>
#include <algorithm>

using namespace cv;
using namespace std;

// Comparison function for contour areas
static bool contour_compare(vector<Point> contour1, vector<Point> contour2)
{
  return contourArea(contour1) < contourArea(contour2);
}

/*
 * Function for detecting a red balloon in an image by finding red areas.
 * Returns a bool that indicates if there was a match or not.
 * The function also calculates the balloon's offset from the image center
 * and stores them in x_offset and y_offset
 */
bool color_matching(Mat& img, int& x_offset, int& y_offset)
{
  int img_w = img.size().width;
  int img_h = img.size().height;
  bool detected = false;
  
  Mat hsv;

  // Blur image
  GaussianBlur(img, hsv, Size(5, 5), 0);

  // Convert to HSV color space
  cvtColor(hsv, hsv, COLOR_BGR2HSV);

  // Mask out the red areas in the image
  Mat mask1, mask2, mask;
  inRange(hsv, Scalar(0, 120, 120), Scalar(10, 255, 255), mask1);
  inRange(hsv, Scalar(170, 120, 120), Scalar(179, 255, 255), mask2);
  addWeighted(mask1, 1.0, mask2, 1.0, 0.0, mask);

  vector<vector<Point> > contours;
  vector<Vec4i> hierarchy;

  // Find contours of the image
  findContours(mask, contours, hierarchy, RETR_TREE, CHAIN_APPROX_SIMPLE, Point(0, 0));

  if (contours.size() > 0)
    {
      // Find the contour with the largest area, and find the smallest bounding box
      vector<Point> c = * max_element(contours.begin(), contours.end(), contour_compare);
      Rect boundRect = boundingRect(c);

      if (boundRect.area() > 500) // Maybe change the area?
	{
	  detected = true;

	  // Draw a rectangle around the found contour
	  rectangle(img, boundRect.tl(), boundRect.br(), Scalar(255, 0, 0), 3);

	  // Find the offset
	  x_offset = int((boundRect.br().x + boundRect.tl().x - img_w) / 2);
	  y_offset = (img_h - boundRect.tl().y - boundRect.br().y) / 2;
	}
    }

  return detected;
}
int main()
{
  // Create a VideoCapture instance
  VideoCapture cap;

  // Check that camera is found
  if (!cap.open(0))
    {
      cout << "Camera not found" << endl;
      return 0;
    }

  for (;;)
    {
      Mat frame;
      cap >> frame; // Read frame from camera
      
      if (frame.empty())
	break;

      // Create ints to store the offsets
      int x_offset, y_offset;

      // Perform color_matching
      bool found = color_matching(frame, x_offset, y_offset);

      // Show result
      namedWindow("TEST", WINDOW_NORMAL);
      imshow("TEST", frame);

      if (waitKey(20) == 27) break; // stop by pressing ESC
    }

  return 0;
}
